gstd/src/common/reservation_manager.rs

```diff
@@ -0,0 +1,155 @@
+///
+/// extern "C" fn init() {
+///     unsafe {
+///         RESERVATION_MANAGER = Some(ReservationManager::new());
+///
+///         RESERVATION_MANAGER.as_mut().unwrap().reserve(200_000, 50);
+///         RESERVATION_MANAGER.as_mut().unwrap().reserve(100_000, 100);
+///         RESERVATION_MANAGER.as_mut().unwrap().reserve(50_000, 30);
+///     }
+/// }
+///
+/// extern "C" fn handle() {
+///     let reservation = unsafe {
+///         RESERVATION_MANAGER
+///             .as_mut()
+///             .unwrap()
+///             .try_take_reservation(100_000)
+///     };
+///     match reservation {
+///         Some((rid, amount)) => {
+///             msg::send_bytes_from_reservation(
+///                 rid,
+///                 msg::source(),
+///                 b"send_bytes_from_reservation",
+///                 0,
+///             )
+///             .expect("Failed to send message from reservation");
+///         }
+///         None => {
+///             msg::send_bytes(msg::source(), b"send_bytes", 0).expect("Failed to send message");
+///         }
+///     }
+/// }
+/// ```
+///
+/// # See also
+/// - [`ReservationId`](ReservationId) is used to reserve and unreserve gas
+/// for program execution later.
+#[derive(Default)]
+pub struct ReservationManager {
+    reservations: Vec<Reservation>,
+}
+
+impl ReservationManager {
+    /// Create a new ReservationManager
+    pub fn new() -> ReservationManager {
+        ReservationManager {
+            reservations: vec![],
+        }
+    }
+
+    /// Reserve the `amount` of gas for further usage.
+    ///
+    /// `duration` is the block count within which the reserve must be used.
+    pub fn reserve(&mut self, amount: u64, duration: u32) -> Result {
+        let id = ReservationId::reserve(amount, duration)?;
+        let block_height = exec::block_height();
+        let valid_until = duration.saturating_add(block_height);
+
+        let insert_to = self.reservations.partition_point(|reservation| {
+            reservation.amount < amount && reservation.valid_until < valid_until
+        });
+
+        // self.reservations will allocate new memory.
+        if self.reservations.capacity() == self.reservations.len() {
+            self.cleanup(block_height);
+        }
+
+        self.reservations.insert(
+            insert_to,
+            Reservation {
+                id,
+                amount,
+                valid_until,
+            },
+        );
+
+        Ok(())
+    }
+
+    /// Find the appropriate reservation with reserved amount greater than or
+    /// equal to `amount`
+    ///
+    /// If such a reservation is found, [`ReservationId`](ReservationId) and
+    /// reserved amount are returned
+    ///
+    /// # See also
+    /// - [`ReservationId`](ReservationId) is used to reserve and unreserve gas
+    ///   amount
+    /// for program execution later.
+    pub fn try_take_reservation(&mut self, amount: u64) -> Option<(ReservationId, u64)> {
```

Comment:
This funtion must return `Result`, with `Err(less_than_amount)` if there's no value `>= amount`